### PR TITLE
Implement a registry mechanism for plugins

### LINKED ISSUE #130: Implement a registry mechanism in maktaba for extending plugins
Some maktaba plugins like codefmt follow a modular design with a user interface and extensible functionality that can be provided by other plugins. The way maktaba models dependencies, this requires a library that serves as the registry/extension point and a plugin that provides the functionality, in addition to any actual plugins providing extensions. For instance, codefmtlib is an extremely simple library that currently defines three functions:

  * `codefmtlib#AddFormatter({formatter})`
  * `codefmtlib#AddDefaultFormatter({formatter})`
  * `codefmtlib#GetFormatters()`

We plan to continue making plugins extensible like this.

Rather than have two repositories per main plugin, we could abstract some kind of registry interface into maktaba so that codefmtlib and similar libraries would just be absorbed into maktaba. API TBD, but we could probably use the plugin name as (part of) a unique key, and the validation would be just configured when that plugin loads and validate existing entries retrospectively rather than being baked into the library.

### PR DESCRIPTION
**Pull request by malcolmr**
Create a way for plugins to register extensions (arbitrary optionally-validated dicts) against other plugins, avoiding the need to split out small library plugins just for registration (e.g. `vim-codefmtlib`) .

Fixes #130.

### CODE REVIEW

**Comment by malcolmr** [comment]
@dbarnett (or any other reviewer): There are some open questions that I'd like comments on; search for TODO in the source/tests.

I assume we'll want to squash this into a single commit before merging, but I've left it as-is for now.

**Comment by malcolmr** [comment]
Hmm, interesting. I'm not entirely sure why the tests pass when I run them locally, but fail on the CI build. I'm using Vim 7.4.622; the CI build appears to be using 7.3.429 - is that relevant?

**Comment by dbarnett** [comment]
Woohoo, we're getting a lot of benefit out of testing lately! Makes me want to set up a matrix of old and new vim versions to verify we're compatible with all the vim versions we intend to support.

It makes sense that passing script-local function references around would be unreliable. I have no idea why it didn't fail until the last commit. I would move the _is_external check into the GetExtensions and SetValidator functions. Switching to an `@private` autoload function would also probably fix it.

**Comment by malcolmr** [comment]
@dbarnett PTAL.

(The reason it failed at the last commit only is because it only ran the build when I submitted the pull request.)

File: autoload/maktaba/extension.vim [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    call maktaba#error#Shout(v:exception)
+  endtry
+endfunction
+
+
+""
+" @dict ExtensionRegistry
+" Returns the extensions that have been added to this extension registry.
+"
+" Extensions added by the plugin to its own registry are always returned
```
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: That's reversed from how codefmtlib currently works. Makes sense if you want to use last-wins instead of first-wins registration, but in that case we should explicitly prescribe callers will normally want the last matching item, not the first.

File: autoload/maktaba/extension.vim [hunk 2]
```diff
@@ -0,0 +1,159 @@
+""
+" @dict ExtensionRegistry
+" A registry for the extensions used a single plugin.  Extensions are
+" dictionaries with a plugin-specific interpretation
```
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: When I see "dictionary", I often forget it can be something more like an "object" that can have function fields. It might make it easier to understand its purpose to call out that it can contain functions and other data.

Also including inline code examples here or pointing to the examples in vroom would go a long way, since it's a pretty abstract feature.

File: autoload/maktaba/plugin.vim [hunk 3]
```diff
@@ -885,6 +886,18 @@ function! maktaba#plugin#IsLibrary() dict abort
 endfunction
 
 
+""
+" @dict Plugin
+" Returns the extension registry belonging to this plugin.
```
Comment (dbarnett) [inline autoload/maktaba/plugin.vim]: Does `Returns the @dict(ExtensionRegistry) belonging …` work here?

File: autoload/maktaba/extension.vim [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+"
+" Plugins should use |Plugin.GetExtensionRegistry| to gain access to their own
```
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: `@function(#GetExtensionRegistry)` should get you the same result.

File: autoload/maktaba/extension.vim [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+" extension registry, and @function(maktaba#extension#GetRegistry) to gain
```
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: Should be able to shorten this to `@function(#GetRegistry)`.

File: autoload/maktaba/extension.vim [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+" needed.
+" @throws WrongType if {plugin} is not a string.
+function! maktaba#extension#GetInternalRegistry(plugin) abort
+  call maktaba#ensure#IsString(a:plugin)
+
+  if has_key(s:registries, a:plugin)
+    return s:registries[a:plugin]
+  endif
+
+  " TODO: should this contain the plugin's name?
```
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: Interesting you don't need it yet. I suspect you'd need it at some point for clear logging / error messages, but no need to add it until you actually need to do something with it.

File: autoload/maktaba/extension.vim [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  " TODO: is there a better way to construct a do-nothing _validator?
```
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: I don't think there is a better way right now. I've wanted one before just for tinkering around at runtime with code that expects a function, so it would be a reasonable feature request. But in plugins it's usually less weird to just leave it undefined and guard the callsite with `has_key(l:registry, '_validator')`. Or this is fine if you want to leave it.

File: autoload/maktaba/extension.vim [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+      \ }
+  let s:registries[a:plugin] = l:registry
+  return l:registry
+endfunction
+
+
+" Throws NotImplemented.  Used by the inaccessible functions on the external
+" interface.
+function! s:NotAccessible() abort
+  throw maktaba#error#NotImplemented('Not accessible by external plugins.')
```
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: I'd almost say to just inline this into the 2 places you call it. Certainly doesn't save any typing this way. =)

File: autoload/maktaba/extension.vim [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+" result in the validation error being shouted to the user (and the extension
+" will not be added).
+"
+" @throws WrongType if {extension} is not a dict.
+function! maktaba#extension#AddExtension(extension) dict abort
+  call maktaba#ensure#IsDict(a:extension)
+
+  try
+    call maktaba#function#Call(self._validator, [a:extension])
+    if self._external
```
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: Use `:lockvar!` here to prevent accidental in-place mutations on the individual extensions.

File: autoload/maktaba/extension.vim [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+""
+" Returns an extension registry for the given plugin name.
+" @throws WrongType if {plugin} is not a string.
+function! maktaba#extension#GetRegistry(plugin) abort
+  let l:registry = maktaba#extension#GetInternalRegistry(a:plugin)
+
+  " Return an external version of the extension registry with a more limited
+  " interface.
+  let l:registry = copy(l:registry)  " TODO: probably not worth caching this, right?
```
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: Yeah, caching here doesn't seem likely to make any practical difference.

File: vroom/pluginextensions.vroom [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+  :let g:foo_extension = {'filetype': 'foo', 'command': 'foo-format'}
+  :call g:registry.AddExtension(g:foo_extension)
+
+If the original plugin has registered a validator (as is the case here), the
+validator will be called to verify the extension.  Any errors returned from
+the validator will be shouted to the user (and will prevent the extension
+from being registered), but will not be propagated to the caller.
+
+TODO: would it make sense for AddExtension() to return whether it succeed,
```
Comment (dbarnett) [inline vroom/pluginextensions.vroom]: I had the thought of returning success, but I can't think of a scenario when the caller would need to do anything special for failures. Just means the plugin needs fixin'.

File: vroom/pluginextensions.vroom [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+then?
+
+TODO: should we define what exception the validator should throw (or even
```
Comment (dbarnett) [inline vroom/pluginextensions.vroom]: I think it's fine. I considered only catching `ERROR(.*)` and letting unexpected code errors like "Invalid expression" bubble up normally, but I can't really see that being a big deal, either.

File: vroom/pluginextensions.vroom [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+  :call g:registry.AddExtension(g:baz_extension)
+  :echomsg string(g:registry.GetExtensions())
+  ~ [{'filetype': 'baz', 'command': 'baz-format'},
+  | {'filetype': 'foo', 'command': 'foo-format'},
+  | {'filetype': 'bar', 'command': 'bar-format'}]
+
+There is one important difference here: extensions added via the internal
+interface will always precede those added via the external interface.
+
+TODO: should internally-added extensions have an _is_internal boolean, or is
```
Comment (dbarnett) [inline vroom/pluginextensions.vroom]: You mean like making the dicts introspectable so the host plugin can explicitly recover which ones were internal? We can add that later if we discover it matters, but pretty sure ordering them is enough.

**Comment by dbarnett** [comment]
LGTM. Commented on TODOs and left a few other comments.

File: autoload/maktaba/plugin.vim (hunk 3 above)
Comment (malcolmr) [inline autoload/maktaba/plugin.vim]: Yes, done.

File: autoload/maktaba/extension.vim (hunk 4 above)
Comment (malcolmr) [inline autoload/maktaba/extension.vim]: I want the dict method on Plugin, not maktaba#extensions#GetExtensionRegsitry() (which doesn't exist).  Is there a better way to refer to that?

File: autoload/maktaba/extension.vim (hunk 5 above)
Comment (malcolmr) [inline autoload/maktaba/extension.vim]: Done.

File: autoload/maktaba/extension.vim (hunk 6 above)
Comment (malcolmr) [inline autoload/maktaba/extension.vim]: Removed TODO.

File: autoload/maktaba/extension.vim (hunk 7 above)
Comment (malcolmr) [inline autoload/maktaba/extension.vim]: Left as-is.

File: autoload/maktaba/extension.vim (hunk 8 above)
Comment (malcolmr) [inline autoload/maktaba/extension.vim]: Good point, done.

File: autoload/maktaba/extension.vim (hunk 10 above)
Comment (malcolmr) [inline autoload/maktaba/extension.vim]: Removed TODO.

File: vroom/pluginextensions.vroom (hunk 11 above)
Comment (malcolmr) [inline vroom/pluginextensions.vroom]: Removed TODO.

File: vroom/pluginextensions.vroom (hunk 12 above)
Comment (malcolmr) [inline vroom/pluginextensions.vroom]: Removed TODO.

File: vroom/pluginextensions.vroom (hunk 13 above)
Comment (malcolmr) [inline vroom/pluginextensions.vroom]: You mentioned _is_internal in one of the original comments on #130.  I've removed this TODO for now, but that's where it came from.

File: autoload/maktaba/extension.vim (hunk 4 above)
Comment (dbarnett) [inline autoload/maktaba/extension.vim]: Oh, right. `@function(maktaba#plugin#GetExtensionRegistry)` will get it. It should be translated to the dict spelling in the docs.

Comment (malcolmr) [inline autoload/maktaba/extension.vim]: That didn't, but `@function(Plugin.GetExtensionRegistry)` did (and translates to the right tag name,  too, unlike what I had before).  Thanks!

File: autoload/maktaba/extension.vim (hunk 1 above)
Comment (malcolmr) [inline autoload/maktaba/extension.vim]: Thinking about this some more, I think first-wins makes more sense, as long as we also prepend to the lists, so that a later call to `AddExtension()` would override an earlier one. Will fix.

File: autoload/maktaba/extension.vim (hunk 2 above)
Comment (malcolmr) [inline autoload/maktaba/extension.vim]: Added some introductory text and examples.

File: autoload/maktaba/extension.vim (hunk 9 above)
Comment (malcolmr) [inline autoload/maktaba/extension.vim]: Done. Also added a call to `deepcopy()`, as per codefmtlib.

**Comment by malcolmr** [comment]
Okay, I think all the comments above are addressed, so PTAL.

If we're happy with this, I'd like to squash it into a single commit and merge.

**Comment by dbarnett** [comment]
LGTM!
I didn't see any logging yet, but we could add that as follow-up later. Logging at least each time a new registry instance is created along with the plugin name it's created for would leave some breadcrumbs in the log for troubleshooting things like typos in the plugin name.
